### ISSUE TITLE
feat: Allow workflow for releasing already-built shorebird app binaries 

### ISSUE THREAD

**Issue by michael-golfi** [shorebirdtech/shorebird#2216]
**Description**

We use a CI process which would require separately building IPA and AAB files to verify that the apps build for the app stores, and then optionally releasing them based on internal process.

A release isn't always necessary and allows us to rely on patching alone until we need to make a native change. This workflow would save us a lot of build time so we don't need to build with shorebird and then rebuild to release.

**Requirements**

- [ ] Documentation for building IPA and AAB files without release
- [ ] CLI command for releasing shorebird IPA and AAB files

**Comment by eseidel** [comment]
`shorebird release` currently does two big things.
1. It calls out to `flutter build` on your behalf, using Shorebird's fork of Flutter.
2. Stores the built result in Shorebird's cloud for your later usage during patching.

I'm not sure from the description above how you would like things to work.  But I would love to learn.  Can you describe for me in greater detail how you'd like the system to work?

There isn't currently a good way to build with Shorebird's Flutter without using the `shorebird release` command.  `shorebird release` is designed to be a drop-in replacement for `flutter build`.  Building frequently with `shorebird release` shouldn't be a problem. It's always possible to deleted the archived builds later if not needed.  We don't currently provide an api for deleting builds (outside of manually via console.shorebird.dev), but happy to add one.

**Comment by michael-golfi** [comment]
Thanks @eseidel 

I think I botched my explanation. Seems like I've gotten a couple people asking. Let me try again...

Currently we run a cron scheduled build process to build and upload IPA and AAB files to their respective test tracks in the respective app stores. When we are ready for a public release we promote the releases to the public app stores.

Going forward, we will be releasing less and patching more. Which is forcing us to rethink how we are releasing features since the constraints will change with Shorebird. Since we can patch, we are considering patching on every feature release (eg: upon a PR merge). However, we also don't want to lose the ability to release when we need to make a major/minor release. We are also concerned that we would stop releasing often and therefore when we need to release we might have a broken build. I'm sure it's unlikely but still..

So right now we are thinking of continuing to build often (maybe weekly instead of nightly) so if a build stops working then we have some certainty about when it stopped working. We can then have the opportunity to release manually when we need to. This workflow introduces a new (but relatively minor) issue, we would be building twice. Which is where the original request here comes from.

I hope this makes more sense. Please let me know if you have more questions, or if my concerns shouldn't be valid.

**Comment by eseidel** [comment]
> Currently we run a cron scheduled build process to build and upload IPA and AAB files to their respective test tracks in the respective app stores. When we are ready for a public release we promote the releases to the public app stores.

The default assumption of Shorebird is that one would just update the cron to always use `shorebird release` instead of `flutter build`.

> Since we can patch, we are considering patching on every feature release (eg: upon a PR merge).

Releases and patches are separate for Shorebird.  Releases all use unique build numbers.  Patches use the same build number as an existing release (and create a "diff" between the Dart logic in the release vs. the patch).  It is possible to build a release and a patch from the same sources by using separate commands and controlling the --release-version passed to the `shorebird patch` command.  e.g.
```
cd my_project
shorebird release ios
# upload to store
shorebird patch ios --release-version 1.0.0
# To also send a patch of this same code to the previous release.
```

> We are also concerned that we would stop releasing often and therefore when we need to release we might have a broken build. I'm sure it's unlikely but still..

Releasing and patching use the same `flutter build` commands under the covers.  I don't think it's possible to break `release` but not `patch` or vice-versa, at least at the Shorebird level.  You can think of `shorebird patch` as essentially replacing *all* of the Dart code in your released app.  It just does that replacement smartly, re-using as much of the existing Dart code compiled into the app as possible (both to save patch download size, as well as to execute most efficiently on iOS where code that comes through the App Store is signed and can run on the CPU, where as any updates through code push must run within an interpreter, thus slower).

> This workflow introduces a new (but relatively minor) issue, we would be building twice.

I'm not aware of a way around the constraint of building twice (at least when building once for a release and once to make a patch).  Maybe there are ways we can help you make builds faster to reduce the burden of building twice?

I'm not sure I fully understand the goal yet, but hopefully my answers provide more context.

https://docs.shorebird.dev/concepts/ and https://docs.shorebird.dev/architecture/ may also help.

**Comment by eseidel** [comment]
Speculatively closing for now. Please feel welcome to re-open if you have more to share.
